src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,416 @@
+{- HLINT ignore "Use camelCase" -}
+
+-- | Incremental, in-memory run consruction
+--
+module Database.LSMTree.Internal.Run.Construction (
+    -- * Types
+    KOp
+  , RawPages (..)
+  , RawKey
+  , RawValue (RawValue, EmptyRawValue)
+  , Operation (..)
+  , BlobRef (..)
+    -- * Page serialisation
+  , serialisePage
+    -- * Incremental page construction
+  , PageIntermediate
+  , pageIntermediateIsEmpty
+  , pageIntermediateEmpty
+  , pageIntermediateAddElem
+  , pageIntermediateSingleton
+    -- * Exposed only for testing
+    -- ** BitMap
+  , BitMap (..)
+  , Bit
+  , emptyBitMap
+  , appendBit
+  , boolToBit
+    -- ** CrumbMap
+  , CrumbMap (..)
+  , Crumb
+  , emptyCrumbMap
+  , appendCrumb
+  , boolsToCrumb
+  ) where
+
+import           Control.Exception (assert)
+import           Data.Bits
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Builder as BB
+import           Data.ByteString.Short.Internal (ShortByteString (SBS))
+import           Data.Foldable
+import           Data.Maybe (isJust)
+import           Data.Primitive.ByteArray
+import           Data.Sequence (Seq (..))
+import qualified Data.Sequence as Seq
+import           Data.Word
+import qualified Database.LSMTree.Internal.Serialise as S
+
+{-------------------------------------------------------------------------------
+  Types
+-------------------------------------------------------------------------------}
+
+-- | A k\/op pair with an optional blob reference.
+type KOp = (RawKey, Operation, Maybe BlobRef)
+
+kopKey :: KOp -> RawKey
+kopKey (x,_,_) = x
+
+kopOp :: KOp -> Operation
+kopOp (_,x,_) = x
+
+kopBlobRef :: KOp -> Maybe BlobRef
+kopBlobRef (_,_,x) = x
+
+--
+-- Pages
+--
+
+
+-- | A string of raw bytes representing one or more disk pages.
+--
+-- Typically, this type only represents a single page, unless it is the result
+-- of adding a larger-than-page value to a run.
+--
+-- Note: the string of bytes is not padded to the target disk-page size.
+data RawPages = RawPages {
+      -- | The number of pages that the output bytestring spans
+      rawPageCount :: !Int
+    , rawPageBytes :: !BS.ByteString
+    }
+
+--
+-- Keys
+--
+
+-- | A string of raw bytes representing a key.
+type RawKey = S.SerialisedKey
+
+-- | Size of key in number of bytes.
+sizeofKey :: RawKey -> Int
+sizeofKey (S.SerialisedKey ba) = sizeofByteArray ba
+
+-- | Size of key in number of bytes.
+sizeofKey16 :: RawKey -> Word16
+sizeofKey16 = fromIntegral . sizeofKey
+
+-- | Size of key in number of bytes.
+sizeofKey64 :: RawKey -> Word64
+sizeofKey64 = fromIntegral . sizeofKey
+
+--
+-- Values
+--
+
+-- | A string of raw bytes representing a value.
+--
+-- A consequence of this simple representation is that for values spanning
+-- multiple pages, those pages must be read into contiguous, page-sized buffers.
+-- This might require dynamically growing the pool of buffers.
+data RawValue =
+    UnsafeRawValue !Int !Int ByteArray
+  | EmptyRawValue
```

Comment:
We should note here that raw values point to /existing/ buffers that are going to be reused